tools/validators/instance_validator/validate/entity_instance.py

```diff
@@ -15,293 +15,395 @@
+
     Returns:
-      Returns boolean for validity of entity.
+      Returns boolean for validity of links key, defaulting to True if the
+      key is not present.
     """
-    for req_key in self.required_keys:
-      if req_key not in self.entity.keys():
-        print('Missing required key:', req_key)
-        return False
 
-    if not self._ValidateType():
-      print('Invalid type')
-      return False
+    if self.links is None:
+      return True
 
-    if not self._ValidateLinks(entity_instances):
-      print('Invalid links')
-      return False
+    is_valid = True
 
-    if not self._ValidateTranslation():
-      print('Invalid translation')
-      return False
+    entity_type = universe.GetEntityType(self.namespace, self.type_name)
+    type_fields = entity_type.GetAllFields()
+    found_fields = set()
+
+    for link_inst in self.links:
+      if link_inst.source not in entity_instances.keys():
+        print('Invalid link source entity name: {0}'.format(link_inst.source))
+        is_valid = False
+        continue
+
+      src_entity = entity_instances.get(link_inst.source)
+      src_namespace = src_entity.namespace
+      src_type_name = src_entity.type_name
+      src_entity_type = universe.GetEntityType(src_namespace,
+                                                    src_type_name)
+
+      for source_field, target_field in link_inst.field_map.items():
+        # assumes that the namespace is '' for now
+        if not entity_type.HasField('/' + target_field):
+          print('Invalid link target field: ', target_field)
+          is_valid = False
+          continue
+
+        if not src_entity_type.HasField('/' + source_field):
+          print('Invalid link source field: ', source_field)
+          is_valid = False
+          continue
+
+        found_fields.add('/' + target_field)
 
-    if not self._ValidateConnections():
-      print('Invalid connections')
+        if not self._ValidateLinkUnits(universe, source_field, target_field):
+          is_valid = False
+          continue
+
+        if not self._ValidateLinkStates(universe, source_field, target_field):
+          is_valid = False
+          continue
+
+    for field_name, field in type_fields.items():
+      if not field.optional and field_name not in found_fields:
+        print('Required field {0} is missing from links'.format(field_name))
+        is_valid = False
+
+    return is_valid
+
+
+  def _ValidateLinkUnits(self, universe, source_field, target_field):
+    """Validates that units match between linked source and target fields."""
+
+    source_units = universe.GetUnitsMapByMeasurement(source_field)
+    target_units = universe.GetUnitsMapByMeasurement(target_field)
+    if source_units != target_units:
+      print('Unit mismatch in link from {0} to {1}'\
+            .format(source_field, target_field))
       return False
+    return True
+
+
+  def _ValidateLinkStates(self, universe, source_field, target_field):
```

Comment:
the state is usually on a translation. Rename to ValidateStates?

Reply:
Renamed to ValidateLinkStatesMatch. (See above)